### DISCUSSION TITLE
Support One-phase capacity motor speed control with zero crossing dimming

### DISCUSSION THREAD

**Discussion by stefanbode** [arendst/Tasmota#10312] [Ideas]
Tasmota supports with zero crossing detection the dimming of lamps and some motors. The used technology is Phase controlled modulator  (Phasenanschnittsteuerung).
This kind of regulation does not work with Capacity Motors that create the second magnetic field with a capacitor. For the phase shift full waves are required.

here an article how it can work.
https://bascomforum.de/index.php?thread/23-netzspannung-wellengruppen-steuern-für-kondensator-motor/&postID=152#post152

Who can support with some internal knowledge. E.g how to drop waves based on the dimming factor. I can and will make the changes to the zero cross dimmer functionality.

**Comment by vic42** [comment]
The current implementation of TRIAC dimmers with zero-crossing input was added just this summer by stefanbode. See #8789 and commit https://github.com/arendst/Tasmota/commit/911dc64e99c8943466ffb42b767f8dc08ce9e829 for details. It involves thoroughly following the mains frequency by tweaking PWM timing to match the respective mains frequency and phase at any time. It may be possible to add your proposed function into that scheme. I haven't looked at it in that much detail yet.

If I had to write the code for what you propose I would probably start from scratch and have each zero crossing fire a hardware interrupt. Since the interrupt only occurs 100 to 120 times per second there should be ample time available to update the TRIAC state from your respective set-point since we only need to update TRIAC trigger states every 2nd interrupt (at minimum 16 ms). The timing is much less critical than it would be for edge-cutting dimmers while phase synchronicity is paramount.

My rough assessment: It will likely be the easiest implementation path adding an interrupt-driven driver based on reacting directly on zero-cross decoder input. Even multiple channels would be rather easy to implement as they may share the same zero crossing interrupt provided they are powered from the same phase conductor.

A good example of the interface towards the sliders can be found at https://github.com/arendst/Tasmota/blob/development/tasmota/xlgt_02_my92x1.ino .

The general process of adding drivers is documented here: https://tasmota.github.io/docs/Sensor-API/

**Comment by vic42** [comment]
I didn't realize _who_ was writing the post, sorry Stefan.

The dimmer described at that site functions as follows. The whole dimming is based on omitting full sine wave cycles, synchronized to zero crossing (20 ms time slots for 50 Hz).

* For very small duty cycle values it will set the output to **on** state for one slot of 20 ms (at 50 Hz), beginning and ending on a zero crossing. After the **on** period it will stay **off** for an integer multiple of that time, again ending at a zero crossing. I would actually do the same for very large duty cycle values, just inverted.
* For 50% duty it will be one full slot  **on** and one full slot **off**, always synchronized to zero crossing.
* For values close to 50% the **on** and **off** slots will be distributed following a predefined bit pattern to reduce motor noise while maintaining the duty cycle, i.e. the ratio of **on** and **off** cycles proportional to the dimmer value. These patterns can be kept in a table that translates dimmer values to bit patterns.

I would still implement it as i have outlined in the posting above.

### Does it make sense?

The technically proper solution would be using a VFD instead. This is sort of a _poor man's VFD_ that comes with quite some limitations. It does modulate the power sent into the motor but AC motors heavily depend on the AC frequency as their rotation is supposed to be synchronous to AC frequency. That obviously cannot be achieved by the outlined TRIAC circuit, no matter what trigger timing.

I think the driver is easy to implement so it may be still worthwhile even though it technically remains a makeshift solution to the problem of AC motor speed control.

**Reply by vic42** [reply]
This dimming scheme actually serves another possible application: **Electrical heaters**.
These should as well be switched full-wave, zero-crossing only, and speed doesn't matter at all due to their thermal inertia.

**Reply by DarthMob** [reply]
Hi @stefanbode , @vic42 ,

is it already possible to use full wave packets? 
https://de.wikipedia.org/wiki/Schwingungspaketsteuerung

Background: I'd like to use my excess solar power to drive a 2 kW  water heating element.
However i want to "dimm" it between 0 and 2000W depending on the spare power.

2kW with Phasenanschnittsteuerung however will create some deformed power
https://de.wikipedia.org/wiki/Verzerrungsblindleistung

For 50% load (1KW) i would need to exactly switch 25 full waves (50 Hz in Germany).
Since one wave is 20ms it would be goal to fire every 40ms.

For siwtiching i would use either a Robotdyn Dimmer
https://robotdyn.com/dimmer-module-for-16-24a-600v-high-load-1-channel-3-3v-5v-logic.html
or a simple SSR (would have no ZC detection, but actually does not matter when you fire since it will stay on only until the next zero crossing).

Can you guys help me out here?

Thanks a lot!

**Reply by stefanbode** [reply]
Hi, I did quite a lot of testing in this area. Also with the full waves. At the end the dimming experience is a nightmare. I have stopped all investigations and through away the code. Theoretically it is possible to do this. Nasty hack. But the dimming of the pump does not do what it should do. I now have bought a frequency converter on eBay for 60€. Even if the electrical magnet fields are not suited for 3phase operation(one winding is only auxiliary) it works quite well as long as you are finally ok with  ~66% of peak power. My pump was oversized anyhow. Therefore no big deal. Also expect the acoustic be more rough. Not worse but it sounds different.

**Reply by stefanbode** [reply]
You can ping me on Skype: Stefan.Bode I assume you're German, too.

**Comment by stefanbode** [comment]
I tried to experiment with my pump today and hooked it into the interrupt of the counter. A very picky thing what is allowed in an interrupt and what is not allowed. Anyhow, I was able to work with the MODULO (%) on the counter and do a digitalWrite to a relay. The oscilloscope shows all good and I was able to drop e.g. 3 out of 4 sinus waves. Long story short: The motor just stutters and make small hard movements with any HIGH on the pin. All far away from controlling the speed. Maybe it worked better if there is a load, but I do not expect this. Seems like there is no way around controlling the speed through a frequency changer.

**Reply by vic42** [reply]
I think you will _always_ need to spin up the motor first to its design speed and then you can _possibly_ reduce power from there. But as everybody will agree this is far from speed control. It may work more-or-less for some light load fan or alike.

**Comment by TrDA-hab** [comment]
@stefanbode 
I tested this:
1. https://aliexpress.ru/item/32802025086.html   
2. https://aliexpress.ru/item/32814316955.html

![20210212_220235](https://user-images.githubusercontent.com/56403720/107817922-d666b580-6d87-11eb-8638-94f8dcdeb1a9.jpg)
![20210212_220544](https://user-images.githubusercontent.com/56403720/107817994-f302ed80-6d87-11eb-9441-813198d81769.jpg)

First device = no result, not even engine hum.
Second device = works great !!! Video:
https://youtu.be/OXBktK0o8oM

For your information only !!!
